packages/flutter_tools/lib/src/base/error_handling_file_system.dart

```diff
@@ -0,0 +1,230 @@
+    ErrorHandlingDirectory(fileSystem, delegate);
+
+  @override
+  Link wrapLink(io.Link delegate) =>
+    ErrorHandlingLink(fileSystem, delegate);
+
+  @override
+  Future<File> writeAsBytes(
+    List<int> bytes, {
+    FileMode mode = FileMode.write,
+    bool flush = false,
+  }) async {
+    try {
+      return wrap(await delegate.writeAsBytes(
+        bytes,
+        mode: mode,
+        flush: flush,
+      ));
+    } on FileSystemException catch (e) {
+      _handleException(
+        e,
+        'Flutter failed to write to a file at "${delegate.path}"',
+      );
+      rethrow;
+    }
+  }
+
+  @override
+  void writeAsBytesSync(
+    List<int> bytes, {
+    FileMode mode = FileMode.write,
+    bool flush = false,
+  }) {
+    try {
+      delegate.writeAsBytesSync(bytes, mode: mode, flush: flush);
+    } on FileSystemException catch (e) {
+      _handleException(
+        e,
+        'Flutter failed to write to a file at "${delegate.path}"',
+      );
+      rethrow;
+    }
+  }
+
+  @override
+  Future<File> writeAsString(
+    String contents, {
+    FileMode mode = FileMode.write,
+    Encoding encoding = utf8,
+    bool flush = false,
+  }) async {
+    try {
+      return wrap(await delegate.writeAsString(
+        contents,
+        mode: mode,
+        encoding: encoding,
+        flush: flush,
+      ));
+    } on FileSystemException catch (e) {
+      _handleException(
+        e,
+        'Flutter failed to write to a file at "${delegate.path}"',
+      );
+      rethrow;
+    }
+  }
+
+  @override
+  void writeAsStringSync(
+    String contents, {
+    FileMode mode = FileMode.write,
+    Encoding encoding = utf8,
+    bool flush = false,
+  }) {
+    try {
+      return delegate.writeAsStringSync(
+        contents,
+        mode: mode,
+        encoding: encoding,
+        flush: flush,
+      );
+    } on FileSystemException catch (e) {
+      _handleException(
+        e,
+        'Flutter failed to write to a file at "${delegate.path}"',
+      );
+      rethrow;
+    }
+  }
+
+  void _handleException(FileSystemException e, String message) {
+    if (platform.isWindows) {
+      _handleWindowsException(e, message);
+    }
+  }
+
+  void _handleWindowsException(FileSystemException e, String message) {
+    // From:
+    // https://docs.microsoft.com/en-us/windows/win32/debug/system-error-codes
+    const int kDeviceFull = 112;
+    const int kUserMappedSectionOpened = 1224;
+    final int errorCode = e.osError?.errorCode ?? 0;
+    // Catch errors and bail when:
+    switch (errorCode) {
+      case kDeviceFull:
+        throwToolExit(
+          '$message. The target device is full.'
+          '\n$e\n'
+          'Free up space and try again.',
+        );
+        break;
+      case kUserMappedSectionOpened:
+        throwToolExit(
+          '$message. The file is being used by another program.'
+          '\n$e\n'
+          'Do you have an antivirus program running? '
```

Comment:
This might be a good candidate for a FAQ page on flutter that we can link to. I know that at least on windows 10, defender can't be fully disabled and instead  requires the user to manually exclude certain directories.